applications/ShapeOptimizationApplication/python_scripts/analyzer_internal.py

```diff
@@ -27,9 +29,27 @@
 except ImportError:
     convdiff_response_factory = None
 
-
 import time as timer
 
+class IterationScope:

```

Comment:
I actually like the idea of having a separate folder for each design. Question @sunethwarna @armingeiser : How do you plan to treat the external analyzer then? More precisely, having this iteration scope means that an external analyzer would also need to implement the IterationScope things? So, one would need to copy every time from the internal analyzer to an external analyzer, which not makes the definition of an external solver much more difficult, but also yields to code doubling. Have you thought about his?

Reply:
@dbaumgaertner As far as I think, external analyzers does not need to implement this. This is an internal scoping which is used only if a response requires it. If they don;t need it they don't need to implement it. 

As for the answers to main comments
> does this change the default behaviour of the optimizer (are now always folders created)?

No, it will create the folder structure if required by the response

https://github.com/KratosMultiphysics/Kratos/blob/0766630b840fb1f989e40ffc87ccdd6e5347d942/kratos/python_scripts/response_functions/response_function_interface.py#L59-L60

`IterationScope` will check whether reponse needs the folder strucutre. If a specific response function requires it, then folder structure will only be created for that response only. So we can have reponses with and without folder structure. (Tested it with constraints in the PR #8399 , and it works)

> in line with the first question, do the changes affect the tests? From a quick test runs, it seems not?

It does not change the tests nor change the performance of the tests. It only adds additionaly small overhead for checking `IsEvalautedInFolder` per each iteration (which is really a boolean check and negligigble)

Reply:
Ah, understood. The fact that every response may decide on its own makes sense. This way, also an external analyzer can do whatever he likes.